src/treemodel/schema.js

```diff
@@ -0,0 +1,417 @@
+	 *
+	 * @param {String} type Paths' type. Possible values are `ALLOW` or `DISALLOW`.
+	 * @param {Array.<String>} checkPath Path to check.
+	 * @param {String} [attribute] If set, only paths registered for given attribute will be returned.
+	 * @returns {Boolean} `true` if item has any registered matching path, `false` otherwise.
+	 * @private
+	 */
+	_hasMatchingPath( type, checkPath, attribute ) {
+		const itemPaths = this._getPaths( type, attribute );
+
+		checkPath = checkPath.slice();
+
+		// We check every path registered (possibly with given attribute) in the item.
+		for ( let itemPath of itemPaths ) {
+			// We have one of paths registered in the item.
+
+			// Now we have to check every item name from the path to check.
+			for ( let checkName of checkPath ) {
+				// Every item name is expanded to all names of items that item is extending.
+				// So, if on item path, there is an item that is extended by item from checked path, it will
+				// also be treated as matching.
+				const chain = this._schema._extensionChains.get( checkName );
+
+				// Since our paths have to match in given order, we always check against first item from item path.
+				// So, if item path is: B D E
+				// And checked path is: A B C D E
+				// It will be matching (A won't match, B will match, C won't match, D and E will match)
+				if ( chain.indexOf( itemPath[ 0 ] ) > -1 ) {
+					// Every time we have a match, we remove it from `itemPath` so we can still check against first item.
+					itemPath.shift();
+				}
+			}
+
+			// If `itemPath` has no items it means that we removed all of them, so we matched all of them.
+			// This means that we found a matching path.
+			if ( itemPath.length === 0 ) {
+				return true;
+			}
+		}
+
+		// No matching path found.
+		return false;
+	}
+
+	/**
+	 * Custom toJSON method to solve child-parent circular dependencies.
+	 *
+	 * @returns {Object} Clone of this object with the parent property replaced with its name.
+	 */
+	toJSON() {
+		const json = clone( this );
+
+		// Due to circular references we need to remove parent reference.
+		json._schema = '[treeModel.Schema]';
+
+		return json;
+	}
+}
+
+/**
+ * Schema is a run-time created and modified description of which entity in Tree Model is allowed to be inside another
+ * entity. It is checked to verify whether given action can be preformed on Tree Model or whether Tree Model
+ * is in correct state.
+ *
+ * Schema consist of {@link core.treeModel.SchemaItem schema items}, each describing different entity. Entity can be
+ * either a Tree Model element or an abstract group for similar elements. Entities are represented by names. Names
+ * of special/abstract entities should be prefixed by `$` sign.
+ *
+ * Each entity in Schema may have a set of allow/disallow rules. Every rule describes in which entities given entity
+ * can or cannot be.
+ *
+ * Entities can extend other entities. This mechanism allows for grouping entities under abstract names. Whenever a rule
```

Comment:
This was one of the approaches I've been thinking of after implementing current solution. It is true that it might turn out in more understandable `Schema` internals, but it would be a bit cumbersome for end-users. For each new element you introduce, you will have to create a new class + some would have to be already introduced (the default `$inline` and `$block` items). OTOH, most features will already have multiple files in their packages (feature class, command classes, ui integration, etc.) so one more probably would not make a harm.

So I think that current solution is easy as all you need to do is to write one line of code somewhere inside feature class. It is in the same vein as tree model `Element`s where they are distinguished only by their name.

For sure I will look forward to enhance `Schema`. There are a lot of ideas and I already described some here https://github.com/ckeditor/ckeditor5-design/issues/129#issuecomment-188137346

A wild thought: we could do the extension thing also with Tree Model `Element`s... But right now they are nice and simple. Maybe (extended) `Element` could be coupled with `Schema` by `SchemaItem` (or even without it, but I don't want to put everything in one place). This is just a wild thought though, nothing to discuss now, really.